terminology/src/main/java/au/csiro/pathling/sql/udf/PropertyLanguageUdf.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2023 Commonwealth Scientific and Industrial Research
+ * Organisation (CSIRO) ABN 41 687 119 230.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package au.csiro.pathling.sql.udf;
+
+import static au.csiro.pathling.fhirpath.encoding.CodingEncoding.decode;
+import static au.csiro.pathling.sql.udf.TerminologyUdfHelpers.isValidCoding;
+import static java.util.Objects.nonNull;
+import static java.util.Objects.requireNonNull;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.BOOLEAN;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.CODE;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.CODING;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.DATETIME;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.DECIMAL;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.INTEGER;
+import static org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType.STRING;
+
+import au.csiro.pathling.encoders.datatypes.DecimalCustomCoder;
+import au.csiro.pathling.errors.InvalidUserInputError;
+import au.csiro.pathling.fhirpath.encoding.CodingEncoding;
+import au.csiro.pathling.terminology.TerminologyService;
+import au.csiro.pathling.terminology.TerminologyService.Property;
+import au.csiro.pathling.terminology.TerminologyService.PropertyOrDesignation;
+import au.csiro.pathling.terminology.TerminologyServiceFactory;
+import ca.uhn.fhir.model.api.annotation.DatatypeDef;
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.hl7.fhir.r4.model.Coding;
+import org.hl7.fhir.r4.model.DateTimeType;
+import org.hl7.fhir.r4.model.Enumerations.FHIRDefinedType;
+import org.hl7.fhir.r4.model.PrimitiveType;
+import org.hl7.fhir.r4.model.Type;
+
+/**
+ * The implementation of the 'propertyLanguage()' udf.
+ */
+@Slf4j
+public class PropertyLanguageUdf implements SqlFunction,
```

Comment:
I think there is a bit of duplication of code between this class and `PropertyUdf`.

Perhaps you could make `PropertyUdf` accommodate invocations with and without language, through the use of nullable parameters.

An example of how this can be done is in `TranslateUdf`.